### DISCUSSION TITLE
How to configure the NPM datasource with a custom registry?

### DISCUSSION THREAD

**Discussion by nickstolwijk** [renovatebot/renovate#9508] [Request Help] [answered]
We use renovate on our internal Bitbucket server and Jenkins server and also have a Nexus server to store our own libraries and proxy Central (Maven) and npmregistry (NPM).

We configured Renovate to use our Nexus server by adding the following to our Renovate config.js:
```
    packageRules: [
        {
            'matchDatasources': ['maven'],
            'registryUrls': ['https://nexus.example.com/repository/public/']
        },
        {
            'matchDatasources': ['npm'],
            'registryUrls': ['https://nexus.example/repository/npm-group/']
        }
   ]
```
That seemed to work, however after https://github.com/renovatebot/renovate/pull/9142 it is throwing warnings: 
```
 WARN: Custom datasources are not allowed for this datasource and will be ignored (repository=EX/example)
       "datasource": "npm",
       "registryUrls": ["https://nexus.example.com/repository/npm-group/"]
```
It looks like the Maven datasource supports this behaviour but the NPM datasource states `customRegistrySupport = false`. I think I am missing something but how should I configure the NPM datasource to use our internal registry instead of the default one?

**Comment by rarkins** [comment, marked as answer]
That's a limitation specific to npm. You can achieve the same by adding a `registry=` line to either a `.npmrc` in the repo, or add the following:

```
  npmrc: 'registry=https://nexus.example/repository/npm-group/'
```

**Reply by HenrikPoulsen** [reply]
Sorry to revive this. But is there a specific reason why there is this limitation. I have the use case where I want to use a custom registry to replace npm installs in shell scripts and the like. For example:
`npm install some-tool@1.0.0 --registry https://mycustomregistry`
That is perfectly legal to do with the npm cli, so it essentially does support custom registries. And you could setup a regex manager which parses the custom registry and uses that for the lookup in theory. But it seems like renovate ignores it.
So is there some special reason why we disallow this in this scenario?

**Reply by rarkins** [reply]
Can you create a new discussion and more fully describe your use case? e.g. it sounds like you mean that you want to have a dependency sourced from a custom registry without that registry appearing in any `.npmrc`.

**Reply by HenrikPoulsen** [reply]
Done. Created #12312
